File: root/app/models/schedule.py
from sqlalchemy import (
    CheckConstraint,
    Column,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
)
from sqlalchemy.orm import relationship

from app.core.database import Base


class Group(Base):
    __tablename__ = "groups"

    id = Column(Integer, primary_key=True)
    name = Column(String, index=True)
    course = Column(Integer)
    faculty = Column(String)

    students = relationship("User", back_populates="group", passive_deletes=True)


class Schedule(Base):
    __tablename__ = "schedule"

    id = Column(Integer, primary_key=True)
    group_id = Column(
        Integer, ForeignKey("groups.id", ondelete="CASCADE"), index=True, nullable=False
    )
    subject = Column(String, nullable=False)
    teacher = Column(String)
    room = Column(String)
    weekday = Column(String, index=True, nullable=False)
    start_time = Column(DateTime(timezone=True), index=True, nullable=False)
    end_time = Column(DateTime(timezone=True), index=True, nullable=False)
    parity = Column(String, default="both", index=True)
    lesson_type = Column(String, default=None)

    __table_args__ = (
        CheckConstraint("end_time > start_time", name="ck_schedule_time_order"),
        Index("ix_schedule_group_start_time", "group_id", "start_time"),
    )

File: frontend/src/db/schemas/schedule.ts
import type { RxJsonSchema } from "rxdb"

export interface ScheduleDoc {
  id: string
  group_id: string
  subject: string
  teacher?: string
  room?: string
  building?: string
  weekday: string
  start_time: string
  end_time: string
  parity?: string
  lesson_type?: string
  updated_at?: string
}

export const scheduleSchema: RxJsonSchema<ScheduleDoc> = {
  version: 0,
  primaryKey: "id",
  type: "object",
  properties: {
    id: { type: "string", maxLength: 100 },
    group_id: { type: "string", maxLength: 100 },
    subject: { type: "string" },
    teacher: { type: "string" },
    room: { type: "string" },
    building: { type: "string" },
    weekday: { type: "string", maxLength: 20 },
    start_time: { type: "string" },
    end_time: { type: "string" },
    parity: { type: "string" },
    lesson_type: { type: "string" },
    updated_at: { type: "string" },
  },
  required: ["id", "group_id", "subject", "weekday", "start_time", "end_time"],
  indexes: ["group_id", "weekday", ["group_id", "weekday"]],
}
